geth/peers/topicpool.go

```diff
@@ -37,13 +44,60 @@ type TopicPool struct {
 	mu         sync.RWMutex
 	discWG     sync.WaitGroup
 	consumerWG sync.WaitGroup
-	connected  int
-	peers      map[discv5.NodeID]*peerInfo
 	period     chan time.Duration
 
+	peerPool       map[discv5.NodeID]*peerInfoItem // found but not connected peers
+	peerPoolQueue  peerPriorityQueue               // priority queue to find the most recent peer
+	connectedPeers map[discv5.NodeID]*peerInfo     // currently connected peers
+
 	cache *Cache
 }
 
+func (t *TopicPool) addToPeerPool(peer *peerInfo) {
+	if _, ok := t.peerPool[peer.node.ID]; ok {
+		return
+	}
+	item := &peerInfoItem{peerInfo: peer}
+	t.peerPool[peer.node.ID] = item
+	heap.Push(&t.peerPoolQueue, item)
+}
+
+func (t *TopicPool) removeFromPeerPool(nodeID discv5.NodeID) {
+	peer, ok := t.peerPool[nodeID]
+	if !ok {
+		return
+	}
+	delete(t.peerPool, nodeID)
+	heap.Remove(&t.peerPoolQueue, peer.index)
+}
+
+func (t *TopicPool) popFromPeerPool() *peerInfo {
+	if t.peerPoolQueue.Len() == 0 {
+		return nil
+	}
+	item := heap.Pop(&t.peerPoolQueue).(*peerInfoItem)
+	delete(t.peerPool, item.node.ID)
+	return item.peerInfo
+}
+
+func (t *TopicPool) updatePeerInPool(nodeID discv5.NodeID, time mclock.AbsTime) {
+	peer, ok := t.peerPool[nodeID]
+	if !ok {
+		return
+	}
+	peer.discoveredTime = mclock.Now()
```

Comment:
I wonder why we need to use `mclock` here? Go natively supports monotonic clock since Go 1.9.